src/WallpaperEngine/Render/CContext.cpp

```diff
@@ -87,19 +88,40 @@ void CContext::render ()
 
     if (this->m_viewports.empty () == false)
     {
-        bool firstFrame = true;
-        bool renderFrame = true;
+        static Display* display = XOpenDisplay (nullptr);
         auto cur = this->m_viewports.begin ();
         auto end = this->m_viewports.end ();
 
+        Window root = DefaultRootWindow(display);
+        int windowWidth = 1920, windowHeight = 1080;
+        int fullWidth = DisplayWidth (display, DefaultScreen (display));
+        int fullHeight = DisplayHeight (display, DefaultScreen (display));
+
+        Pixmap pm = XCreatePixmap(display, root, fullWidth, fullHeight, 24);
+        GC gc = XCreateGC(display, pm, 0, NULL);
+        XFillRectangle(display, pm, gc, 0, 0, fullWidth, fullHeight);
```

Comment:
I kinda just assumed that I needed to recreate the pm after every draw to the X window. I timed this and it only takes a few nanoseconds but also tested it without recreating a new pm & gc and it still worked fine, so I agree on making it be created only once and update it later.  We can make it either two static vars or move them to become a class members so we can have functions later to update the screen size. I'm leaning towards making them class members. The XFillRectangle will need to be kept to draw on every frame to clear the pm.

Reply:
Upon further testing, keeping only one Pixmap & GC results in flickering. I will retract the commits and revert to before where the Pixmap & GC are created every frame. I think the issue is caused by using the old pixmap and then because the compositor is a separate processes, it could use the pixmap after we call XFillRectangle but before XPutImage. You can see this if you pull those commits and run picom with it. Without a compositor running there is no issue.